File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/CPP/ConvertSortedListToBST.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        // inorder traversal construct a BST to sorted linked list
        // we are doing the construction in opposite way
        ListNode* node = head;
        int size = 0;
        for( ; node != NULL; size++, node = node->next ) ;
        return sortedListToBSTHelp( head, 0, size-1);
    }
    
    TreeNode *sortedListToBSTHelp(ListNode* &head, int start, int end ){
        if( start > end )   return NULL;
        
        int mid = (start + end ) /2;
        TreeNode* leftChild = sortedListToBSTHelp( head, start, mid-1);
        
         // this head is a global variable since we get it by reference
        TreeNode* parent =new TreeNode( head->val );   
        head= head->next; 
        
        parent->left = leftChild;
        parent->right = sortedListToBSTHelp( head, mid+1, end);
        
        return parent;
    }
};


void main(){
  Soluction test;
  ListNode* head;
  test.sortedListToBST(head);
}

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/ConvertSortedListToBST.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        """ convert sorted list to balanced binary search tree"""
        self._curr_node_ = head
        size = 0
        while head != None:
            size = size + 1
            head = head.next
        
        return self.inOrderTraversal( 1, size )

    def inOrderTraversal(self, start, end):
        if start > end:
            return None

        middle = (start + end)/2

        left_child = self.inOrderTraversal(start, middle - 1)

        curr_node = TreeNode( self._curr_node_.val )
        self._curr_node_ = self._curr_node_.next        # forward list node
        
        right_child = self.inOrderTraversal(middle + 1, end)

        curr_node.left = left_child
        curr_node.right = right_child
        
        return curr_node
